### DISCUSSION TITLE
metrics.npz file to be generated after training a SLEAP model

### DISCUSSION THREAD

**Discussion by ozgurcemozgen06** [talmolab/sleap#1286] [Help!]
Hi all,

I've been utilizing SLEAP for my M.Sc thesis on "Multi-Animal Keypoint and Pose Estimation using Deep Learning Algorithms". During my last usages(training) of SLEAP on some dataset, it does not create "metrics.npz" file under models directory. In my previous usages, no such thing happened but now it starts to not create that metrics file so I cannot reach the evaluation results of the trained model. I directly used "sleap-train" command inside the terminal and am wondering if I am missing something as an argument. What might be the reason for this situation? I would be so glad if you may guide me about it. 

With my kindest regards,
Özgür

**Comment by roomrys** [comment]
Hi @ozgurcemozgen06,

The first reason why the metrics would not be output would be if your training_config.json has `outputs.save_outputs` set to `False` effectively passing over the `evaluate` function that outputs the metrics:
https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/nn/training.py#L947-L955

However, if you are using the CLI, then we should be setting this to `True` for you:
https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/nn/training.py#L1925-L1926

The second reason why the metrics would not be output is if we ran into an error in the `evaluate_model` function, either `ValueError: Unrecognized model type` (which I would also be surprised by):
https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/nn/evals.py#L675-L708

or `Failed to compute metrics` (which might be the most likely):
https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/nn/evals.py#L713-L718

Did you happen to get any messages in the terminal that might help us determine the cause?

Thanks,
Liezl

**Comment by ozgurcemozgen06** [comment]
Hi @roomrys 

At one of my trials, the training just stopped at iteration 7, without any cause and in my last trial I got a "killed" message just after getting "Creating dataset for training". It seems like it cannot complete training process 

Thank you for your support, 
Özgür

**Reply by roomrys** [reply]
Hi @ozgurcemozgen06,

Do you mind posting the terminal logs here. Starting a few lines before the Traceback should be good, but including everything from `Happy SLEAPing!` would be good to have just in case.

Thanks,
Liezl

**Reply by ozgurcemozgen06** [reply]
Hi @roomrys 

I am adding the terminal screenshot of my last trial, where I still cannot run training. 

Thanks in advance,
Özgür

https://user-images.githubusercontent.com/83988545/234589444-7bb1d757-bf98-4e2e-b653-4f38c3606c44.png

**Reply by roomrys** [reply]
Great, thanks for the traceback @ozgurcemozgen06! 

So it looks like there are some incompatibilities either with your GPU driver or with your version of CUDA/cuDNN. 

You can check which version of tensorflow, cuda, and cudnn you have by using the `conda list` command. You can then compare this to the list of [tested build configurations](https://www.tensorflow.org/install/source#gpu) for your OS provided by tensorflow. These should be correct off the bat if you installed SLEAP through either [conda package](https://sleap.ai/installation.html#conda-package) or [conda from source](https://sleap.ai/installation.html#conda-from-source).

If the package versions are compatible, then the second thing I would try is updating your GPU driver.  You can download Nvidia drivers from [the Nvidia download page](https://www.nvidia.com/Download/index.aspx?lang=en-us). To see the type of GPU you have, you can run the `nvidia-smi` command which will yield something like below (where my driver is `NVIDIA RTX A5000`):
```
λ nvidia-smi
Wed Apr 26 10:24:02 2023
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 511.79       Driver Version: 511.79       CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA RTX A5000   WDDM  | 00000000:01:00.0  On |                  Off |
| 30%   37C    P8    25W / 230W |   5894MiB / 24564MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

Let us know if that helps or if you need help.

Thanks,
Liezl

**Reply by roomrys** [reply]
Hi @ozgurcemozgen06,

Just looping back to see if we were able to help or if you are still struggling?

Thanks,
Liezl

**Reply by ozgurcemozgen06** [reply]
Hi @roomrys, 

When I run SLEAP in my local computer, i am getting a different error saying 
ValueError: Cannot load file containing pickled data when allow_pickle=False

Thanks a lot, 
Özgür

**Reply by roomrys** [reply]
Hi @ozgurcemozgen06,

That sounds like it might be a completely different issue. I have a feeling it has to do with the video loading, but can't be certain without the Traceback from your terminal. Do you mind providing the traceback - maybe we can start a new discussion for that topic?

Thanks,
Liezl

**Reply by ozgurcemozgen06** [reply]
Hi @roomrys 

```
Device: /physical_device:GPU:0
         Available: True
        Initalized: False
     Memory growth: True
INFO:sleap.nn.training:
INFO:sleap.nn.training:Initializing trainer...
INFO:sleap.nn.training:Loading training labels from: /home/students/oezgen/sleaponmarss/labels.v001.slp
Traceback (most recent call last):
  File "/home/students/oezgen/miniconda/envs/sleap/bin/sleap-train", line 33, in <module>
    sys.exit(load_entry_point('sleap', 'console_scripts', 'sleap-train')())
  File "/home/students/oezgen/sleap/sleap/nn/training.py", line 2013, in main
    trainer = create_trainer_using_cli(args=args)
  File "/home/students/oezgen/sleap/sleap/nn/training.py", line 2005, in create_trainer_using_cli
    video_search_paths=args.video_paths,
  File "/home/students/oezgen/sleap/sleap/nn/training.py", line 675, in from_config
    with_track_only=is_id_model,
  File "/home/students/oezgen/sleap/sleap/nn/training.py", line 152, in from_config
    with_track_only=with_track_only,
  File "/home/students/oezgen/sleap/sleap/nn/training.py", line 170, in from_labels
    training, search_paths=video_search_paths
  File "/home/students/oezgen/sleap/sleap/io/dataset.py", line 2731, in load_file
    return Labels.load_file(filename, search_paths, match_to=match_to)
  File "/home/students/oezgen/sleap/sleap/io/dataset.py", line 1966, in load_file
    filename, for_object="labels", video_search=video_search, *args, **kwargs
  File "/home/students/oezgen/sleap/sleap/io/format/main.py", line 113, in read
    return disp.read(filename, *args, **kwargs)
  File "/home/students/oezgen/sleap/sleap/io/format/dispatch.py", line 56, in read
    return adaptor.read(file, *args, **kwargs)
  File "/home/students/oezgen/sleap/sleap/io/format/hdf5.py", line 139, in read
    labels = cls.read_headers(file, video_search, match_to)
  File "/home/students/oezgen/sleap/sleap/io/format/hdf5.py", line 124, in read_headers
    labels = labels_json.LabelsJsonAdaptor.from_json_data(dicts, match_to=match_to)
  File "/home/students/oezgen/sleap/sleap/io/format/labels_json.py", line 386, in from_json_data
    videos = Video.cattr().structure(dicts["videos"], List[Video])
  File "/home/students/oezgen/miniconda/envs/sleap/lib/python3.7/site-packages/cattr/converters.py", line 201, in structure
    return self._structure_func.dispatch(cl)(obj, cl)
  File "/home/students/oezgen/miniconda/envs/sleap/lib/python3.7/site-packages/cattr/converters.py", line 336, in _structure_list
    for e in obj
  File "/home/students/oezgen/miniconda/envs/sleap/lib/python3.7/site-packages/cattr/converters.py", line 336, in <listcomp>
    for e in obj
  File "/home/students/oezgen/miniconda/envs/sleap/lib/python3.7/site-packages/cattr/converters.py", line 323, in structure_attrs_fromdict
    dispatch(type_)(val, type_) if type_ is not None else val
  File "/home/students/oezgen/miniconda/envs/sleap/lib/python3.7/site-packages/cattr/converters.py", line 407, in _structure_union
    return self._structure_func.dispatch(cl)(obj, cl)
  File "/home/students/oezgen/sleap/sleap/io/video.py", line 1541, in fixup_video
    return Video.make_specific_backend(cl, x)
  File "/home/students/oezgen/sleap/sleap/io/video.py", line 1523, in make_specific_backend
    return backend_class(**attribute_kwargs)
  File "<attrs generated init sleap.io.video.NumpyVideo>", line 3, in __init__
  File "/home/students/oezgen/sleap/sleap/io/video.py", line 538, in __attrs_post_init__
    self.__data = np.load(self.filename)
  File "/home/students/oezgen/miniconda/envs/sleap/lib/python3.7/site-packages/numpy/lib/npyio.py", line 445, in load
    raise ValueError("Cannot load file containing pickled data "
ValueError: Cannot load file containing pickled data when allow_pickle=False
```

This is what pops up in my terminal as I run sleap-train command 

Thanks,
Özgür

**Reply by roomrys** [reply]
Hi @ozgurcemozgen06,

I opened PR #1327 for the above Traceback - if tests pass it should be merged soon. 

#### Some background on the Traceback above
We have experience the above intermittently on `SingleImageVideo`s backends (if your "video" consists of a series of images). I wonder if your video is also a series of images instead of an mp4 or avi? Basically, we use a library `cattr` that handles restructuring information in the slp back into the data classes used by SLEAP. We were allowing `cattr` to infer the data class by looking at its attributes, but occasionally `cattr` would infer incorrectly and attempt to construct the wrong video backend.

Thanks,
Liezl

**Reply by ozgurcemozgen06** [reply]
Hi @roomrys , 

I do not use video input, I have 15,000 images of frames of interacting mice 

Thanks,
Ozgur
